### PR TITLE
Use macros into src/app/encoder.cpp for cluster related methods

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

PR #2752 add new methods to src/app/encoder.c. The idea was that methods such as encode*Command use BufBound directly. So it easy to generate if needed and each command encode only its parameters using BufBound.
It's not a big deal but I would like this pattern to stay.

 #### Summary of Changes
 * Remove _encodeIdentifyClusterCommand and use BufBound directly

### CODE REVIEW

File: src/app/encoder.cpp [hunk 1]
```diff
@@ -174,35 +173,30 @@ uint16_t encodeReadOnOffCommand(uint8_t * buffer, uint16_t buf_length, uint8_t d
     return result;
 }
 
-uint16_t _encodeIdentifyClusterCommand(uint8_t * buffer, uint16_t buf_length, uint8_t command_id, uint8_t destination_endpoint,
-                                       uint16_t * identify_duration)
-{
+/*
+ * Identify Cluster commands
+ *
+ * Pick cluster id as 0x0003 for now
```
Comment (andy31415) [inline src/app/encoder.cpp:179]: sould 0x0003 have a constant instead?

it seems to show up in comments and both querycomment and query methods.

File: src/app/encoder.cpp [hunk 2]
```diff
@@ -65,8 +65,7 @@ static uint16_t doEncodeApsFrame(BufBound & buf, uint16_t profileID, uint16_t cl
     }
     else
     {
-        result = buf.Fit() ? buf.Written() : 0;
-        CHECK_FRAME_LENGTH(result, "Buffer too small");
+        CHECK_FRAME_LENGTH((result = buf.Fit() ? buf.Written() : 0), "Buffer too small");
```
Comment (bzbarsky-apple) [inline src/app/encoder.cpp:68]: Why this change?  Having a side-effect inside the condition there seems pretty hard to read to me...

File: src/app/encoder.cpp [hunk 3]
```diff
@@ -174,35 +173,30 @@ uint16_t encodeReadOnOffCommand(uint8_t * buffer, uint16_t buf_length, uint8_t d
...
-    {
-        buf.PutLE16(*identify_duration);
-    }
+    uint16_t result = _encodeClusterSpecificCommand(buf, destination_endpoint, 0x0003, 0x00);
+    CHECK_COMMAND_LENGTH(result, "Identify");
 
-    result = buf.Fit() ? buf.Written() : 0;
+    buf.PutLE16(duration);
 
+    CHECK_COMMAND_LENGTH((result = buf.Fit() ? buf.Written() : 0), "Identify");
```
Comment (bzbarsky-apple) [inline src/app/encoder.cpp:190]: Again, I am unconvinced by this "assign in a condition" pattern.

File: src/app/encoder.cpp (hunk 1 above)
Comment (vivien-apple) [reply src/app/encoder.cpp:179]: Done.

File: src/app/encoder.cpp (hunk 3 above)
Comment (vivien-apple) [reply src/app/encoder.cpp]: Reverted.

**Review by mspang — DISMISSED** [review]

**Review by rwalker-apple — DISMISSED** [review]

**Comment by rwalker-apple** [comment]
@BroderickCarlin @saurabhst @hawk248 @jelderton ?

**Comment by vivien-apple** [comment]
I did change the code in order to use macros. There will be a bunch of similar methods (like the ones I will add with https://github.com/project-chip/connectedhomeip/pull/2792/commits/8bdcffbbea3055f1a430d1c4f1e9c10f884abf1a) and it seems to makes sense to use macros instead of duplicating this code again and again.

So I'm re-asking reviews from @rwalker-apple and @mspang to take into account the changes.

**Review by mspang — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]
better

I'd really like to see us moving to something that looks like printf() for this stuff, where the ID, the name, and the format string live together in a const table?

File: src/app/encoder.cpp [hunk 4]
```diff
@@ -141,75 +134,76 @@ uint16_t encodeReadAttributesCommand(uint8_t * buffer, uint16_t buf_length, uint
     return buf.Fit() && CanCastTo<uint16_t>(buf.Written()) ? static_cast<uint16_t>(buf.Written()) : 0;
 }
 
+#define READ_ATTRIBUTES(name, cluster_id)                                                                                            \
+    uint16_t attr_id_count = sizeof(attr_ids) / sizeof(attr_ids[0]);                                                                 \
+    uint16_t result        = encodeReadAttributesCommand(buffer, buf_length, destination_endpoint, 0x0006, attr_ids, attr_id_count); \
+    if (result == 0)                                                                                                                 \
+    {                                                                                                                                \
+        ChipLogError(Zcl, "Error encoding %s command", name);                                                                        \
+        return 0;                                                                                                                    \
```
Comment (BroderickCarlin) [inline src/app/encoder.cpp:143]: this return seems unnecessary

File: src/app/encoder.cpp [hunk 5]
```diff
@@ -141,75 +134,76 @@ uint16_t encodeReadAttributesCommand(uint8_t * buffer, uint16_t buf_length, uint
...
+        ChipLogError(Zcl, "Error encoding %s command", name);                                                                      \
+        return 0;                                                                                                                  \
+    }
+
+#define COMMAND_FOOTER(name)                                                                                                       \
+    result = buf.Fit() && CanCastTo<uint16_t>(buf.Written()) ? static_cast<uint16_t>(buf.Written()) : 0;                           \
+    if (result == 0)                                                                                                               \
+    {                                                                                                                              \
+        ChipLogError(Zcl, "Error encoding %s command", name);                                                                      \
+        return 0;                                                                                                                  \
```
Comment (BroderickCarlin) [inline src/app/encoder.cpp:161]: this return seems unnecessary

**Review by BroderickCarlin — APPROVED** [review]
Couple nits otherwise LGTM
